### PR TITLE
Support analytic data

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Support analytic data for evolutions.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp [hunk 1]
```diff
@@ -66,14 +80,34 @@ struct Initialize {
...
+            using analytic_data_tag = OptionTags::AnalyticDataBase;
+            prim_vars->assign_subset(
+                Parallel::get<analytic_data_tag>(local_cache)
+                    .variables(
+                        inertial_coords,
+                        typename Metavariables::analytic_variables_tags{}));
+            return Parallel::get<analytic_data_tag>(local_cache)
+                .equation_of_state();
+          })(
+          std::integral_constant<
```
Comment (moxcodes) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp]: the first argument can be`detail::has_analytic_solution_alias<Metavariables>`, which is either `true_type` or `false_type` (and below). It seems these are the only places the `_v` version is used, so that can probably be omitted as well.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+struct EvolutionMetavars;
```
Comment (moxcodes) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp]: For our understanding, why is this file necessary? In other cases, it would seem preferable to just forward-declare the single struct.

**Review by moxcodes — COMMENTED** [review]
Group code review with @markscheel and Rob.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp]: This is needed to provide the magic type alias `metavariables` globally. I currently can't remember why it's necessary, but it is needed if `EvolutionMetavars` is a template like in the scalar wave case. Does that help?

Comment (kidder) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp]: It is needed for `cmake/AddSpectreExecutable.cmake` to generate the actual executable.  It is only necessary when the metavariables are a template, but since we expect most systems to be templated on dimension, it was decided to do it this way for all systems rather than support two setups.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp [hunk 3]
```diff
@@ -66,14 +80,34 @@ struct Initialize {
...
       PrimitiveVars primitive_vars{num_grid_points};
-      primitive_vars.assign_subset(Parallel::get<solution_tag>(cache).variables(
-          inertial_coords, initial_time,
-          typename Metavariables::analytic_variables_tags{}));
-
-      auto equation_of_state =
-          Parallel::get<solution_tag>(cache).equation_of_state();
+      auto equation_of_state = make_overloader(
+          [ initial_time, &inertial_coords ](
+              std::true_type /*is_analytic_solution*/, const auto prim_vars,
```
Comment (hrueter) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp]: could you replace `auto` with the `gsl::not_null<PrimitiveVars*>`?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp [hunk 4]
```diff
@@ -66,14 +80,34 @@ struct Initialize {
...
+            using solution_tag = OptionTags::AnalyticSolutionBase;
+            prim_vars->assign_subset(
+                Parallel::get<solution_tag>(local_cache)
+                    .variables(
+                        inertial_coords, initial_time,
+                        typename Metavariables::analytic_variables_tags{}));
+            return Parallel::get<solution_tag>(local_cache).equation_of_state();
+          },
+          [&inertial_coords](std::false_type /*is_analytic_solution*/,
+                             const auto prim_vars,
```
Comment (hrueter) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp]: also replace `auto`?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp [hunk 5]
```diff
@@ -113,10 +147,28 @@ struct Initialize {
           db::get<::Tags::Coordinates<Dim, Frame::Inertial>>(box);
 
       // Set initial data from analytic solution
-      using solution_tag = OptionTags::AnalyticSolutionBase;
       GrVars gr_vars{num_grid_points};
-      gr_vars.assign_subset(Parallel::get<solution_tag>(cache).variables(
-          inertial_coords, initial_time, typename GrVars::tags_list{}));
+      make_overloader(
+          [ initial_time, &inertial_coords ](
+              std::true_type /*is_analytic_solution*/, const auto local_gr_vars,
```
Comment (hrueter) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp]: same here?

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "Domain/DomainCreators/RegisterDerivedWithCharm.cpp"
+#include "Domain/Tags.hpp"
+#include "ErrorHandling/FloatingPointExceptions.hpp"
+#include "Evolution/Actions/ComputeVolumeDuDt.hpp"
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: rename to `ComputeTimeDerivative`

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/PalenzuelaEtAl.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/System.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Tags.hpp"
+#include "Evolution/VariableFixing/Actions.hpp"
+#include "Evolution/VariableFixing/FixToAtmosphere.hpp"
+#include "Evolution/VariableFixing/Tags.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyBoundaryFluxesGlobalTimeStepping.hpp"
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: rename to `ApplyFluxes`

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  using analytic_data_tag = OptionTags::AnalyticData<analytic_data>;
+  // using boundary_condition_tag = analytic_solution_tag;
+  using analytic_variables_tags =
+      typename system::primitive_variables_tag::tags_list;
+  using equation_of_state_tag = hydro::Tags::EquationOfState<
+      typename analytic_data_tag::type::equation_of_state_type>;
+  using normal_dot_numerical_flux = OptionTags::NumericalFluxParams<
+      dg::NumericalFluxes::LocalLaxFriedrichs<system>>;
+  using limiter = OptionTags::SlopeLimiterParams<
+      SlopeLimiters::Minmod<3, system::variables_tag::tags_list>>;
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: retrieve dimension from `system::volume_dim` (also below)

Alternatively comment that the system always assumes 3 dimensions

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                 StepChoosers::Register::Constant,
+                 StepChoosers::Register::Increase>;
+  using ordered_list_of_primitive_recovery_schemes = tmpl::list<
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::NewmanHamlin,
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::PalenzuelaEtAl>;
+
+  using compute_rhs = tmpl::flatten<tmpl::list<
+      Actions::ComputeVolumeFluxes,
+      dg::Actions::SendDataForFluxes<EvolutionMetavars>,
+      Actions::ComputeVolumeSources, Actions::ComputeVolumeDuDt,
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: rename to `ComputeTimeDerivative`

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+      // dg::Actions::ImposeDirichletBoundaryConditions<EvolutionMetavars>,
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: Add a comment on which boundary conditions can be used at the moment.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+      dg::Actions::ReceiveDataForFluxes<EvolutionMetavars>,
+      tmpl::conditional_t<local_time_stepping, tmpl::list<>,
+                          dg::Actions::ApplyBoundaryFluxesGlobalTimeStepping>,
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: rename to `ApplyFluxes`

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+          VariableFixing::Actions::FixVariables<
+              VariableFixing::FixToAtmosphere<thermodynamic_dim>>,
+          Actions::UpdateConservatives, Actions::FinalTime,
+          tmpl::conditional_t<local_time_stepping,
+                              Actions::ChangeStepSize<step_choosers>,
+                              tmpl::list<>>,
+          compute_rhs, update_variables, Actions::Goto<EvolvePhaseStart>>>>>;
+
+  using const_global_cache_tag_list =
+      tmpl::list<analytic_data_tag,
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: This should probably be in `Initialize::const_global_cache_tag_list` similar to how `normal_dot_numerical_fluxes_tag` is handled.
But `DgElementArray` does not read that list from the initialize action. So maybe we should open an issue for this.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                 OptionTags::TypedTimeStepper<tmpl::conditional_t<
+                     local_time_stepping, LtsTimeStepper, TimeStepper>>,
+                 OptionTags::DampingParameter>;
+
+  using domain_creator_tag = OptionTags::DomainCreator<3, Frame::Inertial>;
+
+  static constexpr OptionString help{
+      "Evolve analytic data using the Valencia formulation of the GRMHD system "
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: List the choices of cylindrical blast wave, Local Lax Friedrichs etc. here.

**Review by hrueter — CHANGES_REQUESTED** [review]
group review by @nilsleiffischer @hrueter

**Review by geoffrey4444 — DISMISSED** [review]
Does this blast wave executable differ from my modified executable in https://github.com/geoffrey4444/spectre/commit/2adab7b43318b70c10773bbda72e713b43aeef82
? Just curious, not easy to diff on my phone...

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp (hunk 8 above)
Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: Pretty much all of GRMHD is coded this way since the system is defined only in 3D. I can add a comment.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: I don't know where to add this comment. You can only use periodic BCs for analytic data right now, but that's a restriction completely outside of this file or even this evolution system.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp (hunk 12 above)
Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: I filed issue #1191. Please feel free to comment there if my description of what should be done is insufficient :)

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp (hunk 10 above)
Comment (geoffrey4444) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp]: So what happens when I run the blast wave with x,y not set to be periodic?

**Comment by nilsdeppe** [comment]
- @geoffrey4444 the biggest difference is you can only apply periodic BCs
- I'm planning on adding a test for the Initialize construct.
- I'm hoping to be able to avoid travis timeouts by speeding up the compilation a bit.

Because of the last 2 points I'm marking this as "in progress".

**Comment by geoffrey4444** [comment]
Thanks @nilsdeppe! Sounds good. Just wondering: why restrict to periodic BCs?

**Comment by nilsdeppe** [comment]
Because we don't have anything else implemented. The dirichlet BCs use the analytic solution, which we don't have if we are using analytic data.

**Comment by hrueter** [comment]
Comment by @nilsleiffischer and @hrueter 
In the current code base the executables are not named after specific solutions. Therefore the executable should be renamed to something like
`EvolveValenciaDivCleanAnalyticData`.

**Comment by nilsdeppe** [comment]
> Comment by @nilsleiffischer and @hrueter
> In the current code base the executables are not named after specific solutions. Therefore the executable should be renamed to something like
> `EvolveValenciaDivCleanAnalyticData`.

Sounds good to me.

**Comment by nilsdeppe** [comment]
@hrueter @nilsleiffischer @moxcodes I addressed the comments. This will timeout on the builds until #1216 is merged but is otherwise ready for review.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      "Evolve analytic data using the Valencia formulation of the GRMHD system "
+      "with divergence cleaning.\n\n"
+      "Analytic data: cylindrical blast wave\n"
+      "Slope limiter: Minmod family\n"
+      "Numerical flux: Local Lax-Friedrichs\n"
+      "Variable fixing: FixConservatives and FixToAtmosphere\n"
+      "Local time-stepping: none\n"
+      "Boundary conditions: only periodic are currently supported\n"};
+
+  enum class Phase { Initialization, Evolve, Exit };
```
Comment (hrueter) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp:146]: Do you have an observe action ready for this, which does not use an analytic solution? If so, can you add it here or in another PR?

**Review by hrueter — COMMENTED** [review]
group review by @nilsleiffischer @hrueter 
squash the fixup

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp (hunk 14 above)
Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp:146]: I do not have such an action written

Comment (nilsvu) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp:146]: It‘ll presumably be just the current grmhd observer with everything that needs an analytic solution deleted, to start with. Out of scope for this PR I suppose, but will be necessary for @geoffrey4444 to rebase the blast wave.

Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp:146]: Yes, I'm not sure what he is observing. I think often you don't actually want to observe all the variables anyway because it's a lot of data for things you might not care about.

**Review by hrueter — DISMISSED** [review]

**Review by kidder — COMMENTED** [review]
squash

**Comment by nilsdeppe** [comment]
Okay, squashed. I'd like to rebase this on #1216 before we merge to make sure Travis is working correctly

**Comment by kidder** [comment]
@nilsdeppe you can rebase

**Comment by nilsdeppe** [comment]
Rebased. :crossed_fingers:

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe are you still planning to add a test for the initializer?

**Comment by nilsdeppe** [comment]
@nilsleiffischer I'd like to defer that so I can focus on implementing more generic boundary conditions first. All the new initialization code needs tests so to do it right will be a fair amount of work.

**Comment by nilsvu** [comment]
yes I know, see #1207 😄 I agree that testing the initialization is out of scope for this PR.

**Comment by kidder** [comment]
@wthrowe please look at

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]
